include/miopen/miopen.h

```diff
@@ -4701,6 +4701,106 @@ miopenReduceTensor(miopenHandle_t handle,
 /** @} */
 // CLOSEOUT TensorReduce DOXYGEN GROUP
 
+/**
+ * An miopenProblem will describe a problem for different miopen operations.
+ * For now, this is only used for convolution, but could be used for other
+ * operators in the future(such as GEMM, Pooling, etc)
+ */
+MIOPEN_DECLARE_OBJECT(miopenProblem);
+
+typedef enum
+{
+    miopenProblemDirectionForward        = 0,
+    miopenProblemDirectionBackward       = 1,
+    miopenProblemDirectionBackwardWeight = 2,
+} miopenProblemDirection_t;
+
+// Enums with no meaningful default value should probably have invalid default for fool-proofing.
+typedef enum
+{
+    miopenTensorNameInvalid  = 0,
+    miopenTensorConvolutionX = 1,
+    miopenTensorConvolutionW = 2,
+    miopenTensorConvolutionY = 3,
+} miopenTensorName_t;
+
+typedef enum
+{
+    miopenSearchResultsOrderByTime   = 0,
+    miopenSearchResultsOrderByMemory = 1,
+} miopenSearchResultsOrder_t;
+
+miopenStatus_t miopenCreateConvProblem(miopenProblem_t* problem,
+                                       miopenConvolutionDescriptor_t operatorDesc,
+                                       miopenProblemDirection_t direction);
+
+miopenStatus_t miopenDestroyProblem(miopenProblem_t problem);
+
+// Universal problem should have named tensor fields rather than ordered.
+// Previously, different directions had different input orders, which may lead to confusion.
+// Implicitly ordered fields would also lead to a whole lot of magical ids all over the internals.
+miopenStatus_t miopenSetProblemTensorDescriptor(miopenProblem_t problem,
+                                                miopenTensorName_t name,
+                                                const miopenTensorDescriptor_t descriptor);
```

Comment:
AFAICS this creates a copy of a tensor descriptor and associates that copy with a problem. It is what we really want? Or it is better to bind tensor descriptor to a problem (pass a pointer)? 

/cc @pfultz2

Reply:
Yes, we want a copy. Otherwise user would have to manage lifetime of a `TensorDescriptor` with a problem and a lot of bad things would happen if they would free it and user the problem afterwards, which would require additional internal complexity to produce a sane error. TensorDescriptor is not a large structure by the way.

Reply:
To reduce the overhead of copying we could store the tensor descriptor data in a impl class, and make all public methods `const` so we can store the impl as `std::share_ptr<const impl>`, so copies are just ref count update. Thats what we do for the 'shape` class in migraphx:

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/include/migraphx/shape.hpp#L232

Reply:
Yes, I was thinking about that as well. Though that is irrelevant from the API perspective and we can add it when we would think that actually matters. That may be planned as a future improvement, but I don't think that would be significant realistically.